File: lab13_17/lab17.cs
namespace LabsAsd;
using System.Collections.Generic;
using System.Text;

public class Lab17
{
    public class Node
    {
        public int Data;
        public Node Left;
        public Node Right;

        public Node(int data)
        {
            Data = data;
            Left = null;
            Right = null;
        }
    }

    public class BinarySearchTree
    {
        public Node Root;

        // 1. Поиск в БДП
        public Node Search(Node root, int key)
        {
            if (root == null || root.Data == key)
                return root;

            if (key < root.Data)
                return Search(root.Left, key);

            return Search(root.Right, key);
        }

        // 2. Добавление в БДП
        public Node Insert(Node root, int key)
        {
            if (root == null) return new Node(key);

            if (key < root.Data)
                root.Left = Insert(root.Left, key);
            else if (key > root.Data)
                root.Right = Insert(root.Right, key);

            return root;
        }

        // 3. Удаление из БДП
        public Node Delete(Node root, int key)
        {
            if (root == null) return root;

            if (key < root.Data)
                root.Left = Delete(root.Left, key);
            else if (key > root.Data)
                root.Right = Delete(root.Right, key);
            else
            {
                // Узел с одним потомком или без них
                if (root.Left == null) return root.Right;
                if (root.Right == null) return root.Left;

                // Узел с двумя потомками: ищем минимальный в правом поддереве
                root.Data = MinValue(root.Right);
                root.Right = Delete(root.Right, root.Data);
            }
            return root;
        }

        private int MinValue(Node root)
        {
            int minv = root.Data;
            while (root.Left != null)
            {
                minv = root.Left.Data;
                root = root.Left;
            }
            return minv;
        }

        // 4. Парсинг линейно-скобочной записи: 8(3(1,6),10)
        public Node Parse(string s, ref int pos)
        {
            if (pos >= s.Length || s[pos] == ')' || s[pos] == ',') return null;

            // Читаем число
            StringBuilder valStr = new StringBuilder();
            while (pos < s.Length && char.IsDigit(s[pos]))
            {
                valStr.Append(s[pos]);
                pos++;
            }

            Node node = new Node(int.Parse(valStr.ToString()));

            if (pos < s.Length && s[pos] == '(')
            {
                pos++; // пропускаем '('
                node.Left = Parse(s, ref pos); // Левый потомок

                if (pos < s.Length && s[pos] == ',')
                {
                    pos++; // пропускаем ','
                    node.Right = Parse(s, ref pos); // Правый потомок
                }

                if (pos < s.Length && s[pos] == ')')
                    pos++; // пропускаем ')'
            }
            return node;
        }

        // 5. Вывод в линейно-скобочном виде
        public string ToBracketString(Node node)
        {
            if (node == null) return "";
            string s = node.Data.ToString();
            if (node.Left != null || node.Right != null)
            {
                s += "(" + ToBracketString(node.Left) + "," + ToBracketString(node.Right) + ")";
            }
            return s;
        }
    }
}

File: lab13_17/lab17.py
#17 lab
class Node:
    def __init__(self, val):
        self.val = val
        self.left = None
        self.right = None

def parse_tree(s):
    def helper():
        nonlocal i
        if i >= len(s) or not s[i].isdigit():
            return None
        num = ''
        while i < len(s) and s[i].isdigit():
            num += s[i]
            i += 1
        node = Node(int(num))
        if i < len(s) and s[i] == '(':
            i += 1
            node.left = helper()
            i += 1  # skip ')'
            if i < len(s) and s[i] == '(':
                i += 1
                node.right = helper()
                i += 1  # skip ')'
        return node
    i = 0
    return helper()

def tree_to_string(node):
    if not node:
        return ''
    left = tree_to_string(node.left)
    right = tree_to_string(node.right)
    if node.left or node.right:
        return f"{node.val}({left})({right})"
    return f"{node.val}"

def search(node, val):
    if not node:
        return False
    if val == node.val:
        return True
    elif val < node.val:
        return search(node.left, val)
    else:
        return search(node.right, val)

def insert(node, val):
    if not node:
        return Node(val)
    if val < node.val:
        node.left = insert(node.left, val)
    elif val > node.val:
        node.right = insert(node.right, val)
    return node

def delete(node, val):
    if not node:
        return None
    if val < node.val:
        node.left = delete(node.left, val)
    elif val > node.val:
        node.right = delete(node.right, val)
    else:
        if not node.left:
            return node.right
        if not node.right:
            return node.left
        # замена на минимальный в правом поддереве
        min_larger = node.right
        while min_larger.left:
            min_larger = min_larger.left
        node.val = min_larger.val
        node.right = delete(node.right, min_larger.val)
    return node

# Ввод дерева
tree_str = input("Введите дерево в линейно-скобочной форме: ").strip()
root = parse_tree(tree_str)

# Меню
while True:
    print("\nМеню: 1-Поиск  2-Добавить  3-Удалить  0-Выход")
    choice = input("Выбор: ").strip()
    if choice == '1':
        val = int(input("Значение для поиска: "))
        print("Найдено" if search(root, val) else "Не найдено")
    elif choice == '2':
        val = int(input("Значение для добавления: "))
        root = insert(root, val)
    elif choice == '3':
        val = int(input("Значение для удаления: "))
        root = delete(root, val)
    elif choice == '0':
        break
    else:
        print("Неверный выбор")

# Вывод дерева
print("Итоговое дерево:", tree_to_string(root))